mbtiles.go

```diff
@@ -47,39 +47,66 @@ func FindMBtiles(path string) ([]string, error) {
 	return filenames, err
 }
 
-// Open opens an MBtiles file for reading, and validates that it has the correct
-// structure.
-func Open(path string) (*MBtiles, error) {
-	// try to open file; fail fast if it doesn't exist
-	stat, err := os.Stat(path)
+// OpenInMemory opens an MBtiles file for reading, and validates that it has the correct
+// structure. Then it loads it to in-memory database. Use this function only with files small enough to be
+// loaded in-memory.
+func OpenInMemory(path string) (*MBtiles, error) {
+	modTime, err := fileModTime(path)
 	if err != nil {
-		if errors.Is(err, os.ErrNotExist) {
-			return nil, fmt.Errorf("path does not exist: %q", path)
-		}
 		return nil, err
 	}
 
-	// there must not be a corresponding *-journal file (tileset is still being created)
-	if _, err := os.Stat(path + "-journal"); err == nil {
-		return nil, fmt.Errorf("refusing to open mbtiles file with associated -journal file (incomplete tileset)")
+	format, tilesize, err := validateAndGetFormatAndSize(path)
+	if err != nil {
+		return nil, err
 	}
 
-	// open a single connection first while we are verifying the database
-	// since there are issues closing out a connection pool on error here
-	con, err := sqlite.OpenConn(path, sqlite.SQLITE_OPEN_READONLY|sqlite.SQLITE_OPEN_NOMUTEX)
-	if con != nil {
-		defer con.Close()
+	srcCon, err := sqlite.OpenConn(path, sqlite.SQLITE_OPEN_READONLY)
```

Comment:
Since you are using a single connection here, I wonder if it would make more sense to keep the handling of the initial connection to validate the database in each of the `Open` functions (instead of moving that connection handling to `validateAndGetFormatAndSize`).  That way you could then just read directly from that connection when writing to the in-memory database.  It would be just a small amount of repeated code but avoid the overhead of repeatedly opening a connection.

Reply:
makes sense, validateAndGet... was deleted and logic moved back to Open functions